<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.247">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Izar de Villasante">

<title>XALD Methylation array analysis - Standard analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

<script src="site_libs/htmlwidgets-1.6.1/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="site_libs/datatables-binding-0.27/datatables.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="site_libs/dt-core-1.12.1/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="site_libs/dt-core-1.12.1/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="site_libs/dt-core-1.12.1/js/jquery.dataTables.min.js"></script>
<script src="site_libs/jszip-1.12.1/jszip.min.js"></script>
<link href="site_libs/dt-ext-buttons-1.12.1/css/buttons.dataTables.min.css" rel="stylesheet">
<script src="site_libs/dt-ext-buttons-1.12.1/js/dataTables.buttons.min.js"></script>
<script src="site_libs/dt-ext-buttons-1.12.1/js/buttons.html5.min.js"></script>
<script src="site_libs/dt-ext-buttons-1.12.1/js/buttons.colVis.min.js"></script>
<script src="site_libs/dt-ext-buttons-1.12.1/js/buttons.print.min.js"></script>
<link href="site_libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="site_libs/nouislider-7.0.10/jquery.nouislider.min.js"></script>
<link href="site_libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="site_libs/selectize-0.12.0/selectize.min.js"></script>
<link href="site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">XALD Methylation array analysis</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">XALD Methylation array analysis</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://www.carrerasresearch.org/en/bioinformatics-unit_130420" title="IJC-BIT" class="quarto-navigation-tool px-1" aria-label="IJC-BIT"><i class="bi bi-globe"></i></a>
    <a href="https://github.com/ijcBIT/XALD_web" title="XALD" class="quarto-navigation-tool px-1" aria-label="XALD"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">XALD epigenic landscape</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pages/standard/standard.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Standard</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./signature.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Alternative approach</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./horvath.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Epigenetic clock</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bibliography</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#standard-analysis" id="toc-standard-analysis" class="nav-link active" data-scroll-target="#standard-analysis">Standard analysis</a>
  <ul class="collapse">
  <li><a href="#workflow" id="toc-workflow" class="nav-link" data-scroll-target="#workflow">Workflow:</a>
  <ul class="collapse">
  <li><a href="#pre-processing" id="toc-pre-processing" class="nav-link" data-scroll-target="#pre-processing">1.Pre-processing:</a></li>
  <li><a href="#differential-methylation-analysis" id="toc-differential-methylation-analysis" class="nav-link" data-scroll-target="#differential-methylation-analysis">2.Differential methylation analysis:</a></li>
  <li><a href="#sec-pathway" id="toc-sec-pathway" class="nav-link" data-scroll-target="#sec-pathway">3. Pathway analysis</a></li>
  </ul></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion">Discussion</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/ijcBIT/XALD_web/edit/main/standard.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/ijcBIT/XALD_web/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Standard analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Izar de Villasante </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="standard-analysis" class="level1">
<h1>Standard analysis</h1>
<section id="workflow" class="level2">
<h2 class="anchored" data-anchor-id="workflow">Workflow:</h2>
<p>The standard workflow consists of three steps: 1. pre-processing (qc, signal normalization, filtering); 2. differential analysis (DMPs,DMRs) and 3. functional analysis of DMR associated genes (GO, pathways)</p>
<section id="pre-processing" class="level3">
<h3 class="anchored" data-anchor-id="pre-processing">1.Pre-processing:</h3>
<section id="data-qc" class="level4">
<h4 class="anchored" data-anchor-id="data-qc">Data QC:</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"R/functions.R"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>ss<span class="ot">&lt;-</span><span class="fu">readRDS</span>(<span class="st">"data/SS.rds"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>ss<span class="sc">$</span>age_grp <span class="ot">&lt;-</span> <span class="st">"Adult"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>ss[Age <span class="sc">&lt;</span> <span class="dv">18</span>, age_grp<span class="sc">:</span><span class="er">=</span><span class="st">"Child"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The following filters are applied:</p>
<ol type="1">
<li>Probes: Only probes with a signal detection p.value &lt; 0.01 for each sample are retained. (P-value calculated with <code>minfi::detectionP()</code>)</li>
<li>Sample: Only samples with at least 90% of the probes detected are retained.</li>
</ol>
<p>After applying these filters one sample was discarded: lnf-21.5, CTL.adult, 44</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>ss <span class="ot">&lt;-</span> ss[Sample_Name <span class="sc">!=</span> <span class="st">"lnf-21.5"</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="background-correction-and-signal-normalization" class="level4">
<h4 class="anchored" data-anchor-id="background-correction-and-signal-normalization">Background correction and signal normalization:</h4>
<p>Raw signal is normalized. The <code>preprocessNoob</code> function from <code>minfi</code> package is used to remove background noise and a visual inspection of the data is done <a href="./standard.html#sec-qc">raw-qc</a>. Signal normalization with ssNoob.</p>
</section>
<section id="filtering" class="level4">
<h4 class="anchored" data-anchor-id="filtering">Filtering:</h4>
<p>Known cross-reactive probes and SNPs are filtered out.</p>
</section>
</section>
<section id="differential-methylation-analysis" class="level3">
<h3 class="anchored" data-anchor-id="differential-methylation-analysis">2.Differential methylation analysis:</h3>
<p>Differences in methylation between different groups are analysed using linear regression models. For single probes (DMPs) we use the limma package, for regions of adjacent probes we use DMRcate (also linear regression based).</p>
<section id="parameter-tunning" class="level4">
<h4 class="anchored" data-anchor-id="parameter-tunning">Parameter tunning:</h4>
<p>We have modified our standard pipeline in order to apply similar filters as those described in the paper by Steegena et al. <span class="citation" data-cites="Steegenga2014">Steegenga et al. (<a href="#ref-Steegenga2014" role="doc-biblioref">2014</a>)</span> with similar samples.</p>
<ul>
<li><p>∆β &gt; 0.05: Difference in mean beta values between groups of at least 0.05.</p></li>
<li><p>T &gt; 2 : t-statisitc greater than 2</p></li>
<li><p>pval &lt; 0.01: P-value smaller than 0.01 in the linear model.</p></li>
</ul>
<p>To answer the research questions we employ the following models:</p>
</section>
<section id="model-1-m1-datass_adults.rds-0-condition-noob" class="level4">
<h4 class="anchored" data-anchor-id="model-1-m1-datass_adults.rds-0-condition-noob">model 1: (m1 “data/ss_Adults.rds”, “~0 + Condition”, Noob )</h4>
<p>=&gt; Identify methylation change associated with adult XALD.</p>
<p>Only using adult samples we create a variable called <code>Condition</code> , which contains 3 groups:</p>
<ul>
<li>CTL.adult: the adult controls</li>
<li>AMN disease without cerebral condition</li>
<li>cAMN disease withcerebral condition.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ss<span class="sc">$</span>Condition)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
CTL.adult       AMN      cAMN 
       11        52         8 </code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dtable</span>(ss[, .(Sample_Name,Condition)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">

<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-9a2628b389ee1a3f7e6e" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-9a2628b389ee1a3f7e6e">{"x":{"filter":"top","vertical":false,"filterHTML":"<tr>\n  <td><\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"factor\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;CTL.adult&quot;,&quot;AMN&quot;,&quot;cAMN&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n<\/tr>","extensions":["Buttons"],"fillContainer":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71"],["AMN_PE-005-V1","BAB-1/8-M0","DNA AMCS","DNA CTL-25.2","DNA CTL-35.3","DNA CTL-44.2","DNA CTL-6.5","DNA CTL-61.1","DNA CTL-64.2","DNA CTL-80.1","DNA JCG","DNA JOG","DNA RCP","DNA SPG-102.2","DNA SPG-122.2","E4 EC-007-V1","E4 EC-008-V1","E4 EC-011-V1","E4 EC-013-V1","E4 EC-311-V2","E4 EC-313-V2","E4 PIOBCN-003-M0","EC-001-V4 JOG","EC-002-V1","EC-004-V1","EC-005-V1","EC-006-V1","EC-009-V1","EC-010-V1","EC-0309-M0 es V2","EC-0312-V2","EC-0401-M0","EC-0402-M0","EC-0403-M0","EC-0404-M0","EC-0405-M0","EC-0406-M0","EC-0407-M0","EC-0408-M0","EC-0409-M0","EC-0410-M0","EC-0411-M0","EC-0412-M0","EC-0413-M0","EC-0414-M0","EC-0415-M0","EC-0416-M0","EC-0417-M0","EC-0418-M0","EC-0419-M0","EC-0420-M0","EC-0421-M0","EC-0422-M0","EC-0423-M0","EC-0424-M0","FVF","ILB","JMGM -1","JMGP","JRD","MLB","MVG_post24h","PIOBCN-001-M0","PIOBCN-006-M0-1_de_3","PIOBCN007","PIODON-003-M0","PIODON-004-M0","RCP-2","SGR_post24h","spg-98.2","spg-98.3"],["AMN","AMN","cAMN","CTL.adult","CTL.adult","CTL.adult","CTL.adult","CTL.adult","CTL.adult","CTL.adult","cAMN","cAMN","cAMN","CTL.adult","CTL.adult","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","cAMN","AMN","cAMN","cAMN","cAMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","CTL.adult","CTL.adult"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sample_Name<\/th>\n      <th>Condition<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"paging":true,"pageLength":20,"searching":true,"fixedColumns":true,"autoWidth":false,"scrollX":true,"digits":4,"ordering":true,"dom":"Bfrtip","buttons":[{"extend":"collection","text":"download entire dataset","action":"function ( e, dt, node, config ) {\n                                                    Shiny.setInputValue('test', true, {priority: 'event'});\n                                                    }"},"copy","csv","excel"],"class":"display","server":true,"columnDefs":[{"orderable":false,"targets":0}],"order":[],"orderClasses":false,"orderCellsTop":true,"lengthMenu":[10,20,25,50,100]}},"evals":["options.buttons.0.action"],"jsHooks":[]}</script>
</div>
</div>
<p>Formula: meth ~ 0 + Condition</p>
<p>RESULTS:</p>
<p>No difference at probe-level or region level observed. (pval =?, FDR =?, methylation difference =)</p>
</section>
<section id="model-2-m2-datass.rds-0-condition-noob" class="level4">
<h4 class="anchored" data-anchor-id="model-2-m2-datass.rds-0-condition-noob">model 2: (m2 “data/SS.rds”, “~0 + Condition”, Noob )</h4>
<p>=&gt; Are there any methylation changes associated with any of the experimental groups such as age or disease phenotype?</p>
<p>Combining children and adults we create a <code>Condition</code> variable, which contains 5 groups formed by all available combinations of disease and age phenotypes:</p>
<ul>
<li><p>CTL.adult: Control samples for adults</p></li>
<li><p>CTL.child: Control samples for childs</p></li>
<li><p>AMN: Adults with the disease condition 1 (?check if this really was included?)</p></li>
<li><p>cAMN: Adults with the disease condition 1 &amp; condition 2</p></li>
<li><p>cALD: Children with the disease condition 2</p></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dtable</span>(ss[, .(Sample_Name,Condition)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">

<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-61e7ad2f9e67d6a8cbb2" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-61e7ad2f9e67d6a8cbb2">{"x":{"filter":"top","vertical":false,"filterHTML":"<tr>\n  <td><\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"factor\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;CTL.adult&quot;,&quot;AMN&quot;,&quot;cAMN&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n<\/tr>","extensions":["Buttons"],"fillContainer":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71"],["AMN_PE-005-V1","BAB-1/8-M0","DNA AMCS","DNA CTL-25.2","DNA CTL-35.3","DNA CTL-44.2","DNA CTL-6.5","DNA CTL-61.1","DNA CTL-64.2","DNA CTL-80.1","DNA JCG","DNA JOG","DNA RCP","DNA SPG-102.2","DNA SPG-122.2","E4 EC-007-V1","E4 EC-008-V1","E4 EC-011-V1","E4 EC-013-V1","E4 EC-311-V2","E4 EC-313-V2","E4 PIOBCN-003-M0","EC-001-V4 JOG","EC-002-V1","EC-004-V1","EC-005-V1","EC-006-V1","EC-009-V1","EC-010-V1","EC-0309-M0 es V2","EC-0312-V2","EC-0401-M0","EC-0402-M0","EC-0403-M0","EC-0404-M0","EC-0405-M0","EC-0406-M0","EC-0407-M0","EC-0408-M0","EC-0409-M0","EC-0410-M0","EC-0411-M0","EC-0412-M0","EC-0413-M0","EC-0414-M0","EC-0415-M0","EC-0416-M0","EC-0417-M0","EC-0418-M0","EC-0419-M0","EC-0420-M0","EC-0421-M0","EC-0422-M0","EC-0423-M0","EC-0424-M0","FVF","ILB","JMGM -1","JMGP","JRD","MLB","MVG_post24h","PIOBCN-001-M0","PIOBCN-006-M0-1_de_3","PIOBCN007","PIODON-003-M0","PIODON-004-M0","RCP-2","SGR_post24h","spg-98.2","spg-98.3"],["AMN","AMN","cAMN","CTL.adult","CTL.adult","CTL.adult","CTL.adult","CTL.adult","CTL.adult","CTL.adult","cAMN","cAMN","cAMN","CTL.adult","CTL.adult","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","cAMN","AMN","cAMN","cAMN","cAMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","AMN","CTL.adult","CTL.adult"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sample_Name<\/th>\n      <th>Condition<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"paging":true,"pageLength":20,"searching":true,"fixedColumns":true,"autoWidth":false,"scrollX":true,"digits":4,"ordering":true,"dom":"Bfrtip","buttons":[{"extend":"collection","text":"download entire dataset","action":"function ( e, dt, node, config ) {\n                                                    Shiny.setInputValue('test', true, {priority: 'event'});\n                                                    }"},"copy","csv","excel"],"class":"display","server":true,"columnDefs":[{"orderable":false,"targets":0}],"order":[],"orderClasses":false,"orderCellsTop":true,"lengthMenu":[10,20,25,50,100]}},"evals":["options.buttons.0.action"],"jsHooks":[]}</script>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ss[, Condition])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
CTL.adult       AMN      cAMN 
       11        52         8 </code></pre>
</div>
</div>
<p>Formula: meth ~ 0 + Condition</p>
<p>RESULTS</p>
<p><strong>This model was statistically significant (pval =?) for the comparison (contrast) cALD-cAMN identifying 116 differentially methylated probes (DMPs) (methylation difference =).</strong></p>
<p>However, no differentially methylated regions (DMRs) could be identified (FDR, meth difference ?).</p>
</section>
<section id="model-3-m4-datass.rds-0-condition-age-noob" class="level4">
<h4 class="anchored" data-anchor-id="model-3-m4-datass.rds-0-condition-age-noob">model 3 (m4 “data/SS.rds”, “~0 + Condition + Age”, Noob )</h4>
<p>=&gt; Aside from age (0-82 years) related, are there any specific methylation changes associated with any of the experimental groups?</p>
<p>model 3 = m2 + Age as a covariate<br>
Formula: meth ~ 0 + Condition + Age</p>
<p>No difference at probe-level or region level observed.</p>
</section>
</section>
<section id="sec-pathway" class="level3">
<h3 class="anchored" data-anchor-id="sec-pathway">3. Pathway analysis</h3>
<p>The 116 DMPs (XX hypo, XX hyper) between cALD and cAMN (model 2) were associated with XXX and XXX genes, respectively. For those genes we performed gene set enrichment analysis:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="data/pathways_m2_camd-cald.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Pathway analysis from genes associated to DMPs in model 2.</figcaption><p></p>
</figure>
</div>
</section>
</section>
<section id="discussion" class="level2">
<h2 class="anchored" data-anchor-id="discussion">Discussion</h2>
<p>We performed a <a href="./standard.html#sec-workflow">Standard methylation analysis</a>, following our standard analysis workflow.</p>
<p>Within the adult group we don’t detect significant difference between controls and disease. Incorporating children samples, we observe few differences between children and adults with cerebral ALD/AMN but none in the other age~phenotype comparisons. Correcting for age (as a continous variable) in the analysis, confirms age as the dominating factors.</p>
<p>Nevertheless, the amount of changes related to age in XALD where much less than expected from previous studies (communication with colaborator).</p>
<p>For this reason (and to further study the pattern of age related changes) we propose alternative approaches:</p>
<ol type="1">
<li>Genereate a <a href="./signature.html#sec-residuals">regression model with age-corrected</a> beta values, to remove the effect of age from the methylation values.</li>
<li>Look for specific age signature in our data by predicting <a href="./signature.html#sec-inhouseref">age-related sites within the controls</a> and checking for enrichment of those sites in the disease.</li>
<li>Analysing <a href="./signature.html#sec-PBMC">previously reported age-related sites</a> from literature.<br>
</li>
<li>Reproducing <a href="./horvath.html">age-acceleration analysis</a> from existing literature.</li>
</ol>



</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Steegenga2014" class="csl-entry" role="listitem">
Steegenga, Wilma T., Mark V. Boekschoten, Carolien Lute, Guido J. Hooiveld, Philip J. De Groot, Tiffany J. Morris, Andrew E. Teschendorff, Lee M. Butcher, Stephan Beck, and Michael Müller. 2014. <span>“Genome-Wide Age-Related Changes in DNA Methylation and Gene Expression in Human PBMCs.”</span> <em>Age</em> 36: 1523–40. <a href="https://doi.org/10.1007/S11357-014-9648-X">https://doi.org/10.1007/S11357-014-9648-X</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">© CC-By IJC-Bit, 2023</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This page is built with ❤️ and <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>



</body></html>