---
title: "Summary"
author: "Izar de Villasante"
---

# Summary

In summary, in this project we tryed to answer the following questions:

1. Is there an age-accelerating factor for XALD patients according to the changes in methylation?

2. Which is the methylation mark associated with XALD disease?

In order to give an answer we started by trying the [conventional analysis]() following the standard workflow including preprocessing, data filtering, DMPs & DMRs identification and finally pathway analysis. We tried different combinations of cutoff values and thresholds for each step and we also tried adding more controls from external sources. 

1. Conventional analysis using BIT pipline.

2. Conventional approach + external controls from GEO.

From this first approach we could identify some [enriched pathways](standard.qmd#sec-pathway) with genes associated with DMPs (individual probes). Nevertheless, by adding age to the model as a covariate this differences where lost.  

![Pathway analysis from genes associated to DMPs in model 2 (meth \~ 0 +
Condition).](data/pathways_m2_camd-cald.png){alt="Pathway analysis from genes associated to DMPs in model 2."}


Since the results from this conventional analysis where not satisfying we then tryed some other analysis replicating existing literature on DNA methylation associated with age:  

3. Checking for age-related signatures following the workflow described in @Steegenga2014

Here we saw some differences in methylation for our data for the set related with no change in gene expression.

```{r}
#| label: tbl-adult_z-test_residuals_model
#| tbl-cap: "Adult Disease vs Adult Controls methylation distributions significance test."
readRDS("data/ztab.rds")
```


~~4. Checking age acceleration as decribed by @Horvath2013~~


We identified a particular set of age-related genes that were differentially methylated for the XALD + cerebral condition in adult patients for specific sites that had been previously reported to be related with age.

See @tbl-adult_z-test_cov_model


2. We also identified the main pathways associated with the disease.


